src/main/java/sparksoniq/jsoniq/runtime/iterator/operational/CastableIterator.java

```diff
@@ -0,0 +1,69 @@
+package sparksoniq.jsoniq.runtime.iterator.operational;
+
+import org.rumbledb.api.Item;
+import sparksoniq.exceptions.IteratorFlowException;
+import sparksoniq.exceptions.NonAtomicKeyException;
+import sparksoniq.exceptions.CastableException;
+import sparksoniq.jsoniq.compiler.translator.expr.operational.base.OperationalExpressionBase;
+import sparksoniq.jsoniq.item.AtomicItem;
+import sparksoniq.jsoniq.item.ItemFactory;
+import sparksoniq.jsoniq.runtime.iterator.RuntimeIterator;
+import sparksoniq.jsoniq.runtime.iterator.operational.base.UnaryOperationBaseIterator;
+import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
+import sparksoniq.semantics.types.AtomicType;
+import sparksoniq.semantics.types.AtomicTypes;
+import sparksoniq.semantics.types.ItemTypes;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class CastableIterator extends UnaryOperationBaseIterator {
+    private static final long serialVersionUID = 1L;
+    private final AtomicType _atomicType;
+
+    public CastableIterator(RuntimeIterator child, AtomicType atomicType, IteratorMetadata iteratorMetadata) {
+        super(child, OperationalExpressionBase.Operator.CASTABLE, iteratorMetadata);
+        this._atomicType = atomicType;
+    }
+
+    @Override
+    public Item next() {
+        if (this._hasNext) {
+            List<Item> items = new ArrayList<>();
+            _child.open(_currentDynamicContext);
+            while (_child.hasNext())
+                items.add(_child.next());
```

Comment:
We can stop after adding two values. This is enough to know what to return or throw an error and will not crash if there is a long sequence.